appserver/ejb/ejb-container/src/main/java/com/sun/ejb/containers/StatefulSessionContainer.java

```diff
@@ -302,33 +336,28 @@
     private Method afterBeginMethod;
     private Method beforeCompletionMethod;
     private Method afterCompletionMethod;
-    private boolean isPassivationCapable;
+    private final boolean isPassivationCapable;
 
     /*
      * Cache for keeping ref count for shared extended entity manager.
      * The key in this map is the physical entity manager
      */
 
-    private static final Map<EntityManager, EEMRefInfo> extendedEMReferenceCountMap
-            = new HashMap<EntityManager, EEMRefInfo>();
+    private static final Map<EntityManager, EEMRefInfo> extendedEMReferenceCountMap = new HashMap<>();
+    private static final Map<EEMRefInfoKey, EntityManager> eemKey2EEMMap = new HashMap<>();
```

Comment:
Concurrent maps are not silver bullet. But I too wonder how this module can work safely for so many years. Except it swallows exceptions when there is already some processed. From Java 7 it can be added as suppressed so we don't lose information.
I will not change it until I would have proof it is broken. JMH + TestContainers will do the job. Or maybe JMH with JUnit and some unit test only. Maybe.